src/sparse/impl/KokkosSparse_par_ilut_numeric_impl.hpp

```diff
@@ -0,0 +1,1048 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOSSPARSE_IMPL_PAR_ILUT_NUMERIC_HPP_
+#define KOKKOSSPARSE_IMPL_PAR_ILUT_NUMERIC_HPP_
+
+/// \file KokkosSparse_par_ilut_numeric_impl.hpp
+/// \brief Implementation(s) of the numeric phase of sparse parallel ILUT.
+
+#include <KokkosKernels_config.h>
+#include <Kokkos_ArithTraits.hpp>
+#include <KokkosSparse_par_ilut_handle.hpp>
+#include <KokkosSparse_spgemm.hpp>
+#include <KokkosSparse_spadd.hpp>
+#include <KokkosSparse_Utils.hpp>
+#include <KokkosSparse_SortCrs.hpp>
+
+#include <limits>
+
+//#define NUMERIC_OUTPUT_INFO
+
+namespace KokkosSparse {
+namespace Impl {
+namespace Experimental {
+
+/**
+ * prefix_sum: Take a row_map of counts and transform it to sums
+ */
+template <class IlutHandle, class RowMapType, class PrefixSumView>
+typename IlutHandle::size_type prefix_sum(IlutHandle& ih, RowMapType& row_map, PrefixSumView& prefix_sum_view)
```

Comment:
A prefix sum is a parallel_scan pattern, so this whole function can be done as one efficient parallel kernel. It can also be done in-place so you don't need both row_map and prefix_sum_view.

In this case, we have a utility already so you could replace this function with something like:
```
KokkosKernels::Impl::exclusive_parallel_prefix_sum<RowMapType, typename IlutHandle::HandleExecSpace
  (row_map.extent(0), row_map);
```

Reply:
Actually, since you need to get the final sum over all elements, you have to do something like this. It will populate total_sum with the final count, as if it were a reduce.

```
    typename RowMapType::non_const_value_type total_sum;
    Kokkos::parallel_scan(Kokkos::RangePolicy<Device>(0, row_map.extent(0)),
        KOKKOS_LAMBDA(size_t i, size_t& lsum, bool finalPass)
        {
          local_ordinal_type count = 0;
          if(i < numAggs)
            count = row_map(i);
          if(finalPass)
            row_map(i) = lsum;
          lsum += count;
        }, total_sum);
```